### PR TITLE
Request executor schema disposal changes and test

### PR DESCRIPTION
**Pull request by matt-psaltis**
See #4863

### CODE REVIEW

File: src/HotChocolate/Stitching/src/Stitching.Redis/RedisExecutorOptionsProvider.cs [hunk 1]
```diff
@@ -105,7 +105,8 @@ await GetRemoteSchemaDefinitionAsync(schemaName)
             IList<IConfigureRequestExecutorSetup> factoryOptions,
             CancellationToken cancellationToken)
         {
-            await using ServiceProvider services =
+            // Don't dispose of the Service Provider here. Causes ObjectDisposedExceptions in RequestExecutorResolver.BeginRunEvictionEvents!
+            ServiceProvider services =
```
Comment (michaelstaib) [inline src/HotChocolate/Stitching/src/Stitching.Redis/RedisExecutorOptionsProvider.cs:109]: Where do we handle disposing the service provider?

**Comment by michaelstaib** [comment]
I changed the target to main ... we first get it into main and then do a cherry-pick to v12

**Comment by michaelstaib** [comment]
Although in this case we do it the other way around since we would have to many merge issues.

File: src/HotChocolate/Stitching/src/Stitching.Redis/RedisExecutorOptionsProvider.cs (hunk 1 above)
Comment (matt-psaltis) [reply src/HotChocolate/Stitching/src/Stitching.Redis/RedisExecutorOptionsProvider.cs:109]: Call chain is:
https://github.com/ChilliCream/hotchocolate/blob/437e97a5764092c6a75b50b93876e010196e91db/src/HotChocolate/Core/src/Execution/RequestExecutorResolver.cs#L162

https://github.com/ChilliCream/hotchocolate/blob/437e97a5764092c6a75b50b93876e010196e91db/src/HotChocolate/Core/src/Execution/RequestExecutorResolver.cs#L461

This is currently prevented as the RegisteredExecutor.Dispose() is never called due to accessing the prematurely disposed  ServiceProvider during the Eviction Actions.

**Comment by matt-psaltis** [comment]
@michaelstaib Any suggestions on removing the 5 minute delay? It makes the integration test problematic. https://github.com/ChilliCream/hotchocolate/blob/437e97a5764092c6a75b50b93876e010196e91db/src/HotChocolate/Core/src/Execution/RequestExecutorResolver.cs#L161

I was thinking about trying to use the diagnostics events to track in flight requests - thoughts?

**Comment by michaelstaib** [comment]
Yes ... we could track if there are still requests in the executor ... on the executor, we could keep a number `ActiveRequests` with `Interlocked.Increment`. This would allow us to remove the fixed time and rather check against this count ... once the count has reached 0 or a grace period is over we will evict.

**Comment by matt-psaltis** [comment]
> Yes ... we could track if there are still requests in the executor ... on the executor, we could keep a number `ActiveRequests` with `Interlocked.Increment`. This would allow us to remove the fixed time and rather check against this count ... once the count has reached 0 or a grace period is over we will evict.

I've had a go at this, not sure if I've accounted for Deferred and Batch queries correctly but I think so?

**Comment by michaelstaib** [comment]
#4874

**Comment by matt-psaltis** [comment]
> #4874

Hi Michael what was the context for that link sorry. Was that for the active request tracking?

**Comment by michaelstaib** [comment]
@matt-psaltis the active count would not work like done. I have reworked the result objects to be simpler and allow release tracking. We now have two result objects, `QueryResult` and `ResponseStream`. A `ResponseStream` represents a stream of `QueryResult`, we use this one now for subscriptions, defer, stream and batch. With each result you can register a cleanup task which will be triggered when the result is disposed. Also I am introducing a completed event on the `ResponseStream`. This completed event is what we can subscribe to in this PR. 

In any case, we need to lift this PR on main since the new APIs will only be available in V13.

**Pull request closed without merging**